### DISCUSSION TITLE
Drop Ruby < 3.0 support

### DISCUSSION THREAD

**Discussion by AlexWayfer** [lostisland/faraday#1543] [Ideas / feature requests]
## Basic Info

* Faraday Version: 2.8.1
* Ruby Version: 2.7.8

## Issue description

1. Ruby <= 2.7 is no longer supported, even with security patches.
2. Now simple standard `gem update --system` or `gem install bundler` are broken, but I believe there no reason to report about this, because, as I've already wrote, it's no longer supported.

## Steps to reproduce

1. Install Ruby 2.7.8.
2. Run `gem update --system` or `gem install bundler`.

## Details

```
> gem update --system
Updating rubygems-update
Fetching rubygems-update-3.5.3.gem
ERROR:  Error installing rubygems-update:
	There are no versions of rubygems-update (= 3.5.3) compatible with your Ruby & RubyGems
	rubygems-update requires Ruby version >= 3.0.0. The current ruby version is 2.7.8.225.
ERROR:  While executing gem ... (NoMethodError)
    undefined method `version' for nil:NilClass
```

https://cirrus-ci.com/task/4917812326563840?logs=install#L0

```
> gem install bundler
ERROR:  Error installing bundler:
	The last version of bundler (>= 0) to support your Ruby & RubyGems was 2.4.22. Try installing it with `gem install bundler -v 2.4.22`
	bundler requires Ruby version >= 3.0.0. The current ruby version is 2.7.8.225.
```

https://cirrus-ci.com/task/6537682302533632?logs=install#L0

---

What do you think? Do you really still want to continue to support such old Ruby versions?

**Comment by AlexWayfer** [comment]
Probably it should be a discussion. :thinking: Sorry, I'm still not familiar with this separation, you're free to move/convert.

**Reply by iMacTia** [reply]
No worries, it's easy to convert issues into discussion 🙌

**Comment by iMacTia** [comment]
Thank you for raising this @AlexWayfer, you're absolutely right in that our current aim in Ruby support is stated in the README:

> This library aims to support and is [tested against](https://github.com/lostisland/faraday/actions) the currently officially supported Ruby implementations. This means that, even without a major release, we could add or drop support for Ruby versions, following their [EOL](https://endoflife.date/ruby).

That means (despite what the outdated README says in the next sentence), that we should support Ruby 3.0+, at least until March when 3.0 will reach EOL and we will be able to drop it.

In practice though, this is a "fallback" to protect maintainers from unnecessary and focus our limited time and resources on meaningful things. If an EOL'ed Ruby version is still working fine or could be supported, then I see no problem with keeping it around a while longer.

However, if there's any particular advantage in dropping an older version or that unlocks improvements that will benefit the wider audience, that will justify dropping EOL'ed rubies from the list of supported ones.

In this particular case, the thing that comes to mind is that we still have `ruby2_keywords` as a dependency because of our support of Ruby 2.7, hence dropping that will allow us to remove the dependency. I think that's a strong enough argument to consider it 🙂

**Reply by AlexWayfer** [reply]
> However, if there's any particular advantage in dropping an older version or that unlocks improvements that will benefit the wider audience, that will justify dropping EOL'ed rubies from the list of supported ones.
> 
> In this particular case, the thing that comes to mind is that we still have `ruby2_keywords` as a dependency because of our support of Ruby 2.7, hence dropping that will allow us to remove the dependency. I think that's a strong enough argument to consider it 🙂

Yes, it can be an advantage too.

But please consider my main point. I understand that supporting old versions can be better for users than not supporting them (although they're unsafe), and we have no analytics (statistics) with what Ruby versions is the gem used.

The main point is that even standard (built-in) commands and gems has been stopped working with Ruby <= 2.7. The Bundler and the `gem update --system`. The advantage for me would be easier CI setup, without any work-arounds, like specifying the latest Bundler and `rubygems-update` versions, which were supporting Ruby 2.7 (and 2.6).

I mean… even not considering safety issues with old Ruby versions, when Ruby "breaks itself" — probably, it's time to move forward.

**Reply by iMacTia** [reply]
Currently giving it a try in #1544 👍

**Reply by iMacTia** [reply]
Version 2.9.0 is now out with a minimum Ruby version of 3.0 🎉
